lib/spack/spack/build_systems/autotools.py

```diff
@@ -140,16 +158,46 @@ def patch(self):
 
     def autoreconf(self, spec, prefix):
         """Not needed usually, configure should be already there"""
-        pass
+        # If configure exists nothing needs to be done
+        if os.path.exists(self.configure_abs_path):
+            return
+        # Else try to regenerate it
+        autotools = ['m4', 'autoconf', 'automake', 'libtool']
+        missing = [x for x in autotools if x not in spec]
+        if missing:
+            msg = 'Cannot generate configure: missing dependencies {0}'
+            raise RuntimeError(msg.format(missing))
+        tty.msg('Configure script not found: trying to generate it')
+        tty.warn('*********************************************************')
+        tty.warn('* If the default procedure fails, consider implementing *')
+        tty.warn('*        a custom AUTORECONF phase in the package       *')
+        tty.warn('*********************************************************')
+        with working_dir(self.configure_directory):
+            m = inspect.getmodule(self)
+            m.libtoolize()
+            m.aclocal()
+            m.autoconf()
+            autogen = Executable('./autogen.sh')
+            autogen()
```

Comment:
I know this is how `swiftsim` does things, but is this generic enough for other packages? Most packages that need it seem to run something like:
```python
autoreconf('--install', '--verbose', '--force')
```
Is your method a more robust version of the same thing?

We may want to comment out the `autoreconf` method I added to `AutoreconfPackageTemplate` in `spack create`, just to let users know that there is now a default and it should suffice most of the time. I would be interested in seeing how many of the other packages in Spack that need to generate a configure script could be converted to use the default.